hack/generator/pkg/astmodel/allof_type.go

```diff
@@ -117,7 +117,7 @@ func (allOf AllOfType) AsDeclarations(codeGenerationContext *CodeGenerationConte
 }
 
 // RequiredImports returns the union of the required imports of all the AllOf types
-func (allOf AllOfType) RequiredImports() []PackageReference {
+func (allOf AllOfType) RequiredImports() *PackageImportSet {
```

Comment:
I think you're conflating `PackageReference` and `PackageImport` here (and in other places below where this same substitution was performed).

A `PackageReference` is a link in a dependency graph (Type `A` uses types from package `P`).
A `PackageImport` is a specific, realized import that satisfies a `PackageReference` requirement.

The key difference being that a `PackageImport` can have a name (`myfancypackage` rather than just a reference `github.com/foo/bar/v1`). `PackageImport` was created to only be used at code generation time (You can see they are first constructed in `generateImports` of `file_definition.go`).

These two types could theoretically be merged into one type, but I personally find it clearer as two - one which is about the graph relationship (and thus cannot be named), and one which is about code (and thus can be named).

I think this means either:
1. We should have a `PackageReferenceSet` in addition to a `PackageImportSet`.
2. We should leave `[]PackageReference`'s as is, and only change places where we were already using `map[PackageImport]` to `PackageImportSet`.

Reply:
Or, if we really want to remove the distinction between the two at the type level (and just say that when a package import doesn't have a name then it can also be used in the "graph" context), then we ought to be deleting/combining `PackageImport` and `PackageReference` - but as I mentioned above I'm not a huge fan of that because I think they really are two different things.

Reply:
Not conflating the two, but fixing a problem. The method is used to answer the question _"What imports does this Type need?"_ but the return type wasn't a list of imports, but a list of references. Some types need to have imports with a particular name, and couldn't specify it - so there was special casing for `metav1` elsewhere to compensate. This method isn't used for the reference graph.

Reply:
It's not possible for the type to answer the question: `what imports does this type need` though, because names need to be derived after conflict resolution which needs the picture of all imports in a file (not just a single type).

The way it was supposed to work here was the type specifies a package reference, and then at code-gen time it's named appropriately. We can have a chat about this if you want over teams?